ide/coqide/preferences_ui.ml

```diff
@@ -0,0 +1,690 @@
+    in
+    let i = ref 0 in
+    let cb = ref [] in
+    let iter text tag =
+      (GMisc.label ~text ~packing:(grid#attach ~left:0 ~top:!i) ())#set_halign `START;
+      let button = tag_button () in
+      let callback () = tag#set button#tag in
+      button#set_tag tag#get;
+      grid#attach ~left:1 ~top:!i button#coerce;
+      incr i;
+      cb := callback :: !cb;
+    in
+    Util.String.Map.iter iter (list_tags ());
+    let label = "Tag configuration" in
+    let callback () = List.iter (fun f -> f ()) !cb in
+    { label; box; apply = callback }
+  in
+
+  let modifiers_valid_callbacks = ref [] in
+  let modifiers_reg callback =
+    modifiers_valid_callbacks := callback :: !modifiers_valid_callbacks
+  in
+  let allowed_mods = str_to_mod_list modifiers_valid#get in
+  let modifiers_valid =
+    create_pref_box "Allowed modifiers" [
+        pmodifiers "" all_modifiers ~callback:(fun value ->
+            List.iter (fun callback -> callback value) !modifiers_valid_callbacks;
+          ) modifiers_valid;
+      ]
+  in
+  let config_modifiers =
+    create_pref_box ~in_grid:true "Modifiers for menu item accelerators" [
+        pmodifiers "View:" allowed_mods ~registerer:modifiers_reg modifier_for_display;
+        pmodifiers "Navigation:" allowed_mods ~registerer:modifiers_reg modifier_for_navigation;
+        pmodifiers "Templates:" allowed_mods ~registerer:modifiers_reg modifier_for_templates;
+        pmodifiers "Queries:" allowed_mods ~registerer:modifiers_reg modifier_for_queries;
+        (* user_queries *)
+      ]
+  in
+
+  let cmd_editor =
+    let predefined = ["gedit %s"; "kate %s"; "emacs %s"; "vi %s"; "notepad %s"; "notepad++ %s"] in
+    pstringcombo ~tooltip:"(%s for file name)" "External editor:" ~editable:true
+      ((if List.mem cmd_editor#get predefined then [] else [cmd_editor#get]) @ predefined)
+      cmd_editor
+  in
+  let cmd_browse =
+    let predefined = [Coq_config.browser; "firefox %s &"; "seamonkey %s &"; "chromium %s &"] in
+    pstringcombo ~tooltip:"(%s for url)" "Browser:" ~editable:true
+      ((if List.mem cmd_browse#get predefined then [] else [cmd_browse#get]) @ predefined)
+      cmd_browse
+  in
+  let externals_cmds =
+    create_pref_box ~in_grid:true "Commands for external programs" [
+        ptextexpr "coqidetop:"
+          (match cmd_coqtop#get with | None -> "AUTO" | Some x -> x)
+          (fun s -> cmd_coqtop#set (if s = "AUTO" then None else Some s));
+        pstring "coqc:" cmd_coqc;
+        pstring "make:" cmd_make;
+        pstring "coqmakefile:" cmd_coqmakefile;
+        pstring "coqdoc:" cmd_coqdoc;
+        pstring "Print ps:" cmd_print;
+        cmd_editor;
+        cmd_browse;
+      ]
+  in
+
+  let misc =
+    create_pref_box "Miscellaneous" [
+        pbool "Stop interpreting before the current point" stop_before;
+        pbool "Reset coqtop on tab switch" reset_on_tab_switch;
+      ]
+  in
+
+(*
+  let add_user_query () =
```

Comment:
Is there a reason to keep this commented-out code? If so, maybe add a short comment explaining why.

Reply:
It’s not from me, it comes from preferences.ml, so I don’t know.